crates/kitsune-core/src/activitypub/fetcher.rs

```diff
@@ -88,6 +92,27 @@ pub struct Fetcher {
 }
 
 impl Fetcher {
+    async fn fetch_ap_resource<T>(&self, url: &str) -> Result<T>
+    where
+        T: DeserializeOwned + RdfNode,
+    {
+        let response = self.client.get(url).await?;
+        let content_type = response
+            .headers()
+            .get(CONTENT_TYPE)
+            .map(HeaderValue::to_str)
+            .transpose()?;
+
+        if content_type
+            != Some("application/ld+json; profile=\"https://www.w3.org/ns/activitystreams\"")
+            && content_type != Some("application/activity+json")
```

Comment:
Hmm.. makes sense. I'll take a look at this this evening

Reply:
Implemented something similar to this except that I'm ignoring the possibility that the `profile` parameter could contain multiple space-delimited URLs.  
IIRC it's well defined in the ActivityPub spec that the parameter is only allowed to contain that single URL

Reply:
> IIRC it's well defined in the ActivityPub spec that the parameter is only allowed to contain that single URL

Is that really so? The Recommendation [specifies][REC-activitypub-retrieving-objects] that <q>Servers […] *MUST* present the ActivityStreams object representation in response to `application/ld+json; profile="https://www.w3.org/ns/activitystreams"`</q>. It says <q>in response to</q>, not <q>respond with</q>, which, I think, requires that servers respond to requests with the `Accept` header value of that content type, but doesn't prohibit servers from respond with a `Content-Type` header value with additional profiles, which doesn't change the fact that the content is an Activity Streams document.


[REC-activitypub-retrieving-objects]: <https://www.w3.org/TR/2018/REC-activitypub-20180123/#retrieving-objects>

Reply:
As a side note, in this interpretation, the spec doesn't prohibit servers from omitting the `Content-Type` header completely, which has worried me a bit about the interoperability of the change, and actually, I have done a quick experiment on it somewhere around Sep. 28 (which I forgot to tell you), if you're curious:

```sh
#!/bin/bash

# Lists `Content-Type` header of (arbitrarily chosen) major ActivityPub implementations.

testcases=(
    'Friendica' 'https://forum.friendi.ca/profile/helpers'
    'GNU Social' 'https://gnusocial.net/diogo'
    'Lemmy' 'https://lemmy.world/c/lemmyworld'
    'Mastodon' 'https://mastodon.social/@Mastodon'
    'Misskey' 'https://misskey.io/@notify'
    'PeerTube' 'https://framatube.org/video-channels/joinpeertube'
    'Pixelfed' 'https://pixelfed.social/dansup'
    'Pleroma' 'https://stereophonic.space/users/karolat'
)

headers="$(mktemp)"

for service uri in "${testcases[@]}"; do
    # Not using `HEAD` method because that would make Lemmy return `404`
    curl --compressed -fLSs -D "$headers" -H 'Accept: application/ld+json; profile="https://www.w3.org/ns/activitystreams"' "$uri" > /dev/null
    type="$(grep -Ei '^content-type:' < "$headers" | sed -E 's/^[^:]*: *//')"
    echo "$service"$'\t'"$uri"$'\t'"$type"
done

rm "$headers"
```

And here is the result:

```
Friendica	https://forum.friendi.ca/profile/helpers	application/activity+json
GNU Social	https://gnusocial.net/diogo	application/activity+json
Lemmy	https://lemmy.world/c/lemmyworld	application/activity+json
Mastodon	https://mastodon.social/@Mastodon	application/activity+json; charset=utf-8
Misskey	https://misskey.io/@notify	application/ld+json; profile="https://www.w3.org/ns/activitystreams"; charset=utf-8
PeerTube	https://framatube.org/video-channels/joinpeertube	application/activity+json; charset=utf-8
Pixelfed	https://pixelfed.social/dansup	application/activity+json
Pleroma	https://stereophonic.space/users/karolat	application/activity+json; charset=utf-8
```

So, at least these implementations respond with `Content-Type` as expected (though some implementations include `charset` parameter, which isn't defined by JSON-LD nor Activity Streams spec).

Reply:
That should be fine! We are using the `mime` crate to parse the actual MIME type and compare the essence strings, meaning the `charset` parameter won't break anything.

And I just added support for multiple whitespace-separated profile URLs, so it should now cover all bases.